shell/lib/docker.sh

```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env bash
+# Helper functions for Docker image generation.
+
+DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+
+# shellcheck source=./logging.sh
+source "${DIR}/logging.sh"
+
+# shellcheck source=./bootstrap.sh
+source "${DIR}/bootstrap.sh"
+
+# shellcheck source=./yaml.sh
+source "${DIR}/yaml.sh"
+
+# get_image_field is a helper to return a field from the manifest
+# for a given image. It will return an empty string if the field
+# is not set.
+#
+# Arguments:
+#   $1 - image name
+#   $2 - field name
+#   $3 - type (array or string) (default: string)
+#   $4 - manifest file (default: $MANIFEST)
+get_image_field() {
+  local image="$1"
+  local field="$2"
+
+  # type can be 'array' or 'string'. Array values are
+  # returned as newline separated values, string values
+  # are returned as a single string. A string value can
+  # be strings, ints, bools, etc.
+  local type=${3:-string}
+  local manifest=${4:-$MANIFEST}
+  local filter
+
+  filter="$(yaml_construct_object_filter "$image" "$field")"
+  if [[ $type == "array" ]]; then
+    yaml_get_array "$filter" "$manifest"
+    return
+  elif [[ $type == "string" ]]; then
+    yaml_get_field "$filter" "$manifest"
+    return
+  else
+    error "Unknown type '$type' for get_image_field"
+    fatal "Expected 'array' or 'string'"
+  fi
+}
+
+# Generates Docker CLI arguments for building an image.
+#
+# docker_buildx_args(app_name, version, image, dockerfile[, arch]) -> arg string
+docker_buildx_args() {
+  local app_name="$1"
+  local version="$2"
+  local image="$3"
+  local dockerfile="$4"
+  local arch="$5"
+
+  # Platforms to build this image for, expected format (in YAML):
+  # platforms:
+  #   - linux/arm64
+  #   - linux/amd64
+  #
+  # See buildkit docs: https://github.com/docker/buildx#building-multi-platform-images
+  mapfile -t platforms < <(get_image_field "$image" "platforms" "array")
+  if [[ ${#platforms[@]} -eq 0 ]]; then # no platforms found
+    if [[ -n $arch ]]; then
+      platforms=("linux/${arch}")
```

Comment:
nit: If the $arch is an input from yaml, suggest to validate and only allow for the predefined platforms.

Reply:
I don't *particularly* want to try to validate the value here, since this is an open source repo and external users could use arches that are outside of `amd64`/`arm64` (this [Docker documentation page](https://docs.docker.com/build/building/multi-platform/) has a non-exhaustive list of supported arches).